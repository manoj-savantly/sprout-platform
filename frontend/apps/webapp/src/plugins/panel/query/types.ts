import { QueryParametersConfiguration } from './editors/QueryParametersEditor';
import { CodeTemplateEditorConfiguration } from './editors/TemplateEditor';

export type QueryPanelDisplayType = 'FRAME' | 'RENDER';

export interface QueryPanelOptions {
  url: string;
  queryParameters: QueryParametersConfiguration;
  displayType: QueryPanelDisplayType;
  template: CodeTemplateEditorConfiguration;
  useTemplate: boolean;
  useProxy: boolean;
  headers: string;
}
