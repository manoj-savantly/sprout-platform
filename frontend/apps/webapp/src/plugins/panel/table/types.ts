import { QueryParametersConfiguration } from './editors/QueryParametersEditor';
import { CodeTemplateEditorConfiguration } from './editors/TemplateEditor';

export interface TablePanelOptions {
  url: string;
  queryParameters: QueryParametersConfiguration;
  jsonDataPath: string;
  headerTemplate: CodeTemplateEditorConfiguration;
  useProxy: boolean;
  headers: string;
}
