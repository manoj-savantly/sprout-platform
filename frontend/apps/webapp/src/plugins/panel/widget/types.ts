export interface WidgetPanelOptions {
  dataSourceId: string
  dataId: string
}

export type WidgetDataType = 'MARKDOWN' | 'MARKUP' | 'JSON'

export interface WidgetData {
    id: string;
    name: string;
    dataType: WidgetDataType;
    data: any;
}
