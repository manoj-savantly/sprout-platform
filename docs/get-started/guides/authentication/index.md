---
title: Authentication
description: Learn how to configure authentication for the Sprout UI and Server
---

Sprout supports several authentication mechanisms.  

## Basic  
Basic authentication is provided by default, and an initial admin user is created if you're using the default server.  
You may configure basic auth users to be created at startup using the application properties.  
If the user already exists, the configuration is skipped. This allows you to change a password after provisioning, without it being overwritten.  

Example configuration -  

```
sprout:
  security:
    authentication:
      basic:
        enable: true
        users:
        - username: admin
          password: changeme!
          email-address: example@savantly.net
```

## OAuth2  
You can provide provide the UI with an Implicit OAuth2 flow by configuring the properties.  

Example using keycloak -  
In this example, a realm called `my-realm` and a client called `webapp` have been created in keycloak.  
Implicit flow has also been enabled for the `webapp` client.  

```
sprout:
  security:
    authentication:
      oauth:
        clients:
        - name: keycloak
          display-name: Login with Keycloak Account
          authorization-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/auth
          token-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/token
          user-info-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/userinfo
          jwks-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/certs
          scope: openid profile email
          redirect-url: "http://localhost:8080/"
          clientId: webapp
          flow-type: IMPLICIT
spring:
  security:
    oauth2:
      resourceserver:
        jwt:
          jwk-set-uri: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/certs
```

OAuth can be further configured by providing your own implementations of certain beans.  
Each of these interfaces have default implementations that can be extended, or replaced.  

`OAuth2UserSynchronizer` executes after a successful authentication. It receieves an `OAuth2User` and should produce a `SproutUser`  
`OAuthConfigurer` configures the Spring `HttpSecurity`  
`OAuthUserMapper` maps an `OAuth2User` to a `SproutOAuthUser`  
`OAuth2UserService` receives an `OAuth2UserRequest` and produces an `OAuth2User`.  
`OidcUserService` receives an `OidcUserRequest` and produces an `OAuth2User`.  

## JWT
The backend server application uses standard Spring Authentication properties.  
Example of JWT configuration in the application.properties -  

```
spring:
  security:
    oauth2:
      resourceserver:
        jwt:
          jwk-set-uri: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/certs
```

## Cookie
TODO: Improve Documentation  
After logging in, a session cookie is established to provide authentication when requesting static resources from a static site, such as the Sprout Web UI.  
