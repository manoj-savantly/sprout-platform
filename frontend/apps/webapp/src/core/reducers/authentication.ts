import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { User } from '@savantly/sprout-api';
import axios, { AxiosRequestConfig } from 'axios';
import { ACCESS_TOKEN_STORAGE_KEY, SERVER_API_URL } from '../../config/constants';
import { AuthenticationState, AuthenticationUpdate } from '../../types';
import store from '../store';

export const initialAuthenticationState: AuthenticationState = {
  loading: false,
  isAuthenticated: store.get(ACCESS_TOKEN_STORAGE_KEY) ? true : false,
  loginError: false, // Errors returned from server side
  user: {
    authorities: []
  } as User,
  errorMessage: undefined, // Errors returned from server side
  sessionHasBeenFetched: false,
  sessionFetchFailed: false,
  logoutUrl: undefined,
  showLogin: false
};

export const getSession = createAsyncThunk('authentication/getSession', async (arg, thunkAPI) => {
  const token = store.get(ACCESS_TOKEN_STORAGE_KEY);
  const config: AxiosRequestConfig = {
    headers: {}
  };
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  } else {
    console.log('sending unauthenticated request', config);
  }
  const response = await axios.get(`${SERVER_API_URL}/api/account`, config);
  console.log(response);
  if (response.data) {
    return response.data;
  } else {
    store.delete(ACCESS_TOKEN_STORAGE_KEY);
    return thunkAPI.rejectWithValue(response);
  }
});

const authenticationSlice = createSlice({
  name: 'authentication',
  initialState: initialAuthenticationState,
  reducers: {
    showLogin: (state): AuthenticationState => {
      return {
        ...state,
        showLogin: true
      };
    },
    login: (state, action: PayloadAction<AuthenticationUpdate>): AuthenticationState => {
      const isAuthenticated = action.payload.accessToken ? true : false;
      if (isAuthenticated) {
        store.set(ACCESS_TOKEN_STORAGE_KEY, action.payload.accessToken as string);
      }
      return {
        ...state,
        isAuthenticated,
        errorMessage: action.payload.errorMessage,
        loginError: !isAuthenticated,
        showLogin: !isAuthenticated,
        sessionHasBeenFetched: false,
        sessionFetchFailed: false
      };
    },
    logout: (state): AuthenticationState => {
      if (store.get(ACCESS_TOKEN_STORAGE_KEY)) {
        store.delete(ACCESS_TOKEN_STORAGE_KEY);
      }
      return {
        ...state,
        user: {
          authorities: []
        },
        errorMessage: undefined,
        isAuthenticated: false,
        loading: false,
        loginError: false,
        logoutUrl: undefined,
        showLogin: false
      };
    }
  },
  extraReducers: (builder) => {
    builder.addCase(
      getSession.fulfilled,
      (state, action): AuthenticationState => {
        console.log(action);
        return {
          ...state,
          user: {
            ...state.user,
            ...action.payload
          },
          isAuthenticated: true,
          sessionHasBeenFetched: true,
          sessionFetchFailed: false,
          showLogin: false
        };
      }
    );
    builder.addCase(
      getSession.rejected,
      (state, action): AuthenticationState => {
        console.error(action);
        const authError = !!action.error.message && action.error.message.indexOf('401') > -1;
        return {
          ...state,
          user: {
            ...state.user
          },
          isAuthenticated: false,
          sessionHasBeenFetched: false,
          sessionFetchFailed: true,
          showLogin: authError
        };
      }
    );
  }
});

export const { login, logout, showLogin } = authenticationSlice.actions;

export const authenticationReducer = authenticationSlice.reducer;
