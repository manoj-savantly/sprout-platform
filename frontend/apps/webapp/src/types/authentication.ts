import { User } from '@savantly/sprout-api';

export interface AuthenticationState {
  loading: boolean;
  isAuthenticated: boolean;
  loginError: boolean; // Errors returned from server side
  errorMessage?: string; // Errors returned from server side
  sessionHasBeenFetched: boolean;
  sessionFetchFailed: boolean;
  logoutUrl?: string;
  user: User;
  showLogin: boolean;
}

export interface AuthenticationUpdate {
  errorMessage?: string;
  accessToken?: string;
}

export interface OAuthClientConfig {
  name: string;
  autoLogin?: boolean;
  displayName: string;
  issuerUri: string;
  clientId: string;
  scope: string;
  authorizationUrl: string;
  redirectUrl: string;
  tokenUrl: string;
  userInfoUrl: string;
}