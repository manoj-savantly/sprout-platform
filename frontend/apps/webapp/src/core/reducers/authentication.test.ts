import moxios from 'moxios';
import { AnyAction } from 'redux';
import configureStore, { MockStoreEnhanced } from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AuthenticationState } from '../../types';
import { getSession, initialAuthenticationState } from './authentication';

const middlewares = [thunk]; // add your middlewares like `redux-thunk`
const mockStore = configureStore<{ authentication: AuthenticationState }>(middlewares);

describe('authentication tests', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('has successful login', function (done) {
    const store: MockStoreEnhanced<{ authentication: AuthenticationState }> = mockStore({
      authentication: initialAuthenticationState
    });
    // Match against an exact URL value
    moxios.stubRequest('/api/account', {
      status: 200,
      response: {
        name: 'user@savantly.net',
        authorities: [
          'DASHBOARD_EDIT',
          'FILES_DELETE',
          'FOLDER_CREATE',
          'FOLDER_DELETE',
          'offline_access',
          'FOLDER_READ',
          'FILES_CREATE',
          'DASHBOARD_DELETE',
          'ADMIN',
          'FILES_READ',
          'DASHBOARD_READ',
          'EVERYONE'
        ]
      }
    });

    expect(store.getState().authentication).toStrictEqual(initialAuthenticationState);

    // Return the promise
    store.dispatch((getSession() as unknown) as AnyAction).then(() => {
      const actions = store.getActions();

      expect(actions[0].type).toEqual('authentication/getSession/pending');
      expect(actions[1].type).toEqual('authentication/getSession/fulfilled');

      expect(store.getState()).toMatchSnapshot();
      done();
    });
  });

  it('has failed login', function (done) {
    const store: MockStoreEnhanced<{ authentication: AuthenticationState }> = mockStore({
      authentication: initialAuthenticationState
    });
    // Match against an exact URL value
    moxios.stubRequest('/api/account', {
      status: 401
    });

    expect(store.getState().authentication).toStrictEqual(initialAuthenticationState);

    // Return the promise
    store.dispatch((getSession() as unknown) as AnyAction).then(() => {
      const actions = store.getActions();

      expect(actions[0].type).toEqual('authentication/getSession/pending');
      expect(actions[1].type).toEqual('authentication/getSession/rejected');

      expect(store.getState()).toMatchSnapshot();
      done();
    });
  });
});
