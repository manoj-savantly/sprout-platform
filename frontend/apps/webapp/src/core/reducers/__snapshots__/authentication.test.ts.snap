// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authentication tests has failed login 1`] = `
Object {
  "authentication": Object {
    "errorMessage": undefined,
    "isAuthenticated": false,
    "loading": false,
    "loginError": false,
    "logoutUrl": undefined,
    "sessionFetchFailed": false,
    "sessionHasBeenFetched": false,
    "showLogin": false,
    "user": Object {
      "authorities": Array [],
    },
  },
}
`;

exports[`authentication tests has successful login 1`] = `
Object {
  "authentication": Object {
    "errorMessage": undefined,
    "isAuthenticated": false,
    "loading": false,
    "loginError": false,
    "logoutUrl": undefined,
    "sessionFetchFailed": false,
    "sessionHasBeenFetched": false,
    "showLogin": false,
    "user": Object {
      "authorities": Array [],
    },
  },
}
`;
